### PR TITLE
Feat/climsoft models

### LINKED ISSUE #2: Create SQLAlchemy ORM models for Climsoft CDMS
This issue will begin to add support for the [Climsoft](https://climsoft.org) CDMS. You will need an instance of MySQL 8+ installed and then to restore the Climsoft 4.1.1 SQL DDL file [here](https://github.com/opencdms/opencdms-test-data/tree/main/schemas/climsoft/v4/v4.1.1).

When the Climsoft software is installed, the SQL DDL creates two databases:

    CREATE DATABASE IF NOT EXISTS `mariadb_climsoft_db_v4`
    CREATE DATABASE IF NOT EXISTS `mariadb_climsoft_test_db_v4`

The first is used as the actual climate database whilst the second contains some test data that is used for training with tutorials.

We only need to create models for `mariadb_climsoft_db_v4`. The models should be created in `/opencdms/models/climsoft/v4_1_1.py`.

Once we have the models, we should try using the `get_schema_diff` function from `opencdms.models` to check our new models against the second database schema (`mariadb_climsoft_test_db_v4`) to see whether it is identical.  It's possible that there may be minor differences.

There is one final complication: some of the tables in Climsoft do not have primary keys explicitly specified - this prevents SQLAlchemy/sqlacodegen from creating ORM models (see [docs here](https://docs.sqlalchemy.org/en/14/faq/ormconfiguration.html#how-do-i-map-a-table-that-has-no-primary-key)).  The attached PDF describes where primary keys should be specified.

We should discuss our options for working around the primary key issue.  We can either:
- Submit a change request to the Climsoft project to `ALTER TABLE` and add the missing primary keys, this will then appear in future releases of Climsoft
- Alter our local copy of the schema to make it appear that primary keys are specified.  If my understanding is correct, we should then be able to use our models (that explicitly set the primary keys) to connect to current versions of the database where many primary keys are not explicitly defined, but everything will work correctly.

I'm very happy to discuss this more - either in GitHub comments, messaging or scheduling a call...
[Climsoft Data Model Tables and Keys.pdf](https://github.com/opencdms/pyopencdms/files/6995106/Climsoft.Data.Model.Tables.and.Keys.pdf)

### PR DESCRIPTION
**Pull request by Shaibujnr**
Resolves #2

## TO EXECUTE DATABASE DDL For CLIMSOFT
`mysql -u root -p < script.sql`

## Execute the following statements to add primary key columns to the 24 core tables
```
alter table observationfinal add primary key (recordedFrom,describedBy,obsDatetime, obsLevel);
alter table observationinitial add primary key (recordedFrom,describedBy,obsDatetime,qcStatus,acquisitionType, obsLevel);
alter table instrumentinspection add primary key (performedOn, inspectionDatetime);
alter table faultresolution add primary key (resolvedDatetime, associatedWith);
alter table stationelement add primary key (recordedFrom, describedBy, recordedWith, beginDate);
alter table stationlocationhistory add primary key (belongsTo, openingDatetime);
alter table stationqualifier add primary key (qualifier, qualifierBeginDate, qualifierEndDate, belongsTo);
alter table observationschedule add primary key (classifiedInto, startTime, endTime);
alter table physicalfeature add primary key (associatedWith, beginDate, classifiedInto, description);
alter table featuregeographicalposition add primary key (belongsTo, observedOn);
alter table paperarchive add primary key (belongsTo, formDatetime, classifiedInto);
```

##  GENERATE MODELS
```
sqlacodegen "mysql://root:{{password}}@localhost/mariadb_climsoft_db_v4" --outfile opencdms/models/climsoft/v4_1_1_core.py --tables=observationfinal,observationinitial,station,obselement,instrument,instrumentfaultreport,instrumentinspection,faultresolution,stationelement,stationlocationhistory,stationqualifier,observationschedule,obsscheduleclass,physicalfeature,physicalfeatureclass,featuregeographicalposition,synopfeature,paperarchive,flags,acquisitiontype,qctype,qcstatusdefinition,regkeys,data_forms
```

### CODE REVIEW

**Comment by faysal-ishtiaq** [comment]
I have gone through the PR and it looks good. However, when I followed the same process, there were some dissimilarities in the resultant file. So, I pushed a commit so that we can view the diff. Here is the commit 

https://github.com/opencdms/pyopencdms/commit/1ad710ef35af9b515e9c28d08ad2481d2481e168

Feel free to revert the commit

File: opencdms/models/climsoft/v4_1_1_core.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
+# coding: utf-8
```
Comment (faysal-ishtiaq) [inline opencdms/models/climsoft/v4_1_1_core.py:1]: This file is the result of sqlacodegen run on my machine. I used latest mysql docker image. feel free to revert the commit if necessary. here is the commit https://github.com/opencdms/pyopencdms/commit/1ad710ef35af9b515e9c28d08ad2481d2481e168

**Review by faysal-ishtiaq — COMMENTED** [review]
LGTM. just let me know about the difference in this file  opencdms/models/climsoft/v4_1_1_core.py. See this commit https://github.com/opencdms/pyopencdms/commit/1ad710ef35af9b515e9c28d08ad2481d2481e168 to see the diff

**Comment by isedwards** [comment]
Thank you @faysal-ishtiaq - these are all of the known changes that @Shaibujnr has made manually in order to make the models match the output of `alembic revision --autogenerate` (see the comments from list item 4 onwards [here](https://github.com/opencdms/pyopencdms/issues/2#issuecomment-909061505)).

This is mainly necessary so that SQLAlchemy's doesn't detect a difference and suggest a migration.

Could you revert 1ad710e for us and then I'll merge?

**Comment by faysal-ishtiaq** [comment]
> Thank you @faysal-ishtiaq - these are all of the known changes that @Shaibujnr has made manually in order to make the models match the output of `alembic revision --autogenerate` (see the comments from list item 4 onwards [here](https://github.com/opencdms/pyopencdms/issues/2#issuecomment-909061505)).
> 
> This is mainly necessary so that SQLAlchemy's doesn't detect a difference and suggest a migration.
> 
> Could you revert [1ad710e](https://github.com/opencdms/pyopencdms/commit/1ad710ef35af9b515e9c28d08ad2481d2481e168) for us and then I'll merge?

reverted just now

**Review by faysal-ishtiaq — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -104,3 +104,4 @@ ENV/
 # IDE settings
 .vscode/
 .idea/
+shaibu-activity.log
```
Comment (isedwards) [inline .gitignore]: @faysal-ishtiaq could you remove this for now?
@Shaibujnr - would it be useful to include `*-activity.log` in a future commit? Is this a common file generated by the IDE (or possibly related to upwork?)

File: opencdms/models/__init__.py [hunk 3]
```diff
@@ -55,19 +89,22 @@ def get_schema_diff(
...
+    if include_tables is not None and exclude_tables is not None:
+        raise Exception("`include_tables` and `exclude_tables` must not be used together") #TODO define custom error class
+    if exclude_tables is not None:
+        diff = list(filter(filter_out_excluded_tables(exclude_tables), diff))
+    if include_tables is not None:
+        diff = list(filter(filter_in_included_tables(include_tables), diff))
     return list(diff)
 
 
 # TODO remove and use automated tests instead.
```
Comment (isedwards) [inline opencdms/models/__init__.py]: @faysal-ishtiaq - could you remove this section to the end of the file and propose tests for both clide and climsoft using sqlite instead?

The clide code that was here imported `CLIDE_VIEWS` (`from opencdms.models.clide import metadata, CLIDE_VIEWS`) and used this to exclude these views/tables from the `get_schema_diff()` output.

The new version, that replaces the original clide example, can be turned into a test for climsoft.

**Review by isedwards — CHANGES_REQUESTED** [review]

**Review by isedwards — APPROVED** [review]

**Pull request merged**